engine/crates/engine-v2/src/engine.rs

```diff
@@ -63,30 +63,38 @@ impl Engine {
         }
     }
 
+    pub fn user_hooks(&self) -> &UserHooks {
+        &self.env.user_hooks
+    }
+
     pub async fn execute(
         self: &Arc<Self>,
         headers: http::HeaderMap,
         batch_request: BatchRequest,
     ) -> HttpGraphqlResponse {
-        let headers = match self.env.user_hooks.on_gateway_request(headers).await {
-            Ok(headers) => headers,
+        let (context, headers) = match self.env.user_hooks.on_gateway_request(headers).await {
+            Ok(result) => result,
             Err(error) => return Response::execution_error(error).into(),
         };
 
         if let Some(access_token) = self.auth.authorize(&headers).await {
-            self.execute_with_access_token(RequestMetadata::new(headers, access_token), batch_request)
-                .await
+            let metadata = RequestMetadata::new(headers, access_token, context);
+            self.execute_with_access_token(metadata, batch_request).await
         } else if let Some(streaming_format) = headers.typed_get::<StreamingFormat>() {
             HttpGraphqlResponse::stream_request_error(streaming_format, "Unauthorized")
         } else {
             HttpGraphqlResponse::request_error("Unauthorized")
         }
     }
 
-    pub async fn create_session(self: &Arc<Self>, headers: http::HeaderMap) -> Option<Session> {
+    pub async fn create_session(
+        self: &Arc<Self>,
+        headers: http::HeaderMap,
+        context: HashMap<String, String>,
```

Comment:
`create_session` should call `on_gateway_request`. `Context` is an implementation detail. `create_session` is used in websocket connection to accept multiple incoming requests.